src/renderer/contexts/DialogContext.tsx

```diff
@@ -0,0 +1,91 @@
+import React, { createElement, useCallback, useMemo, useState } from 'react'
+
+import type { JSXElementConstructor, PropsWithChildren } from 'react'
+
+import { generateRandomUUID } from '../utils/random'
+
+type DialogId = string
+
+export type DialogProps = {
+  onClose: () => void
+}
+
+type DialogElementConstructor<T> = JSXElementConstructor<DialogProps & T>
+
+export type OpenDialog = <T extends { [key: string]: any }>(
+  dialogElement: DialogElementConstructor<T>,
+  additionalProps?: T
+) => DialogId
+
+export type CloseDialog = (id: DialogId) => void
+
+type DialogContextValue = {
+  hasOpenDialogs: boolean
+  openDialog: OpenDialog
+  closeDialog: CloseDialog
+}
+
+const initialValues: DialogContextValue = {
+  hasOpenDialogs: false,
+  openDialog: _ => '',
+  closeDialog: _ => {},
+}
+
+export const DialogContext = React.createContext<DialogContextValue>(
+  initialValues
+)
+
+export const DialogContextProvider = ({ children }: PropsWithChildren<{}>) => {
+  const [dialogs, setDialogs] = useState<{ [id: DialogId]: JSX.Element }>({})
+
+  const closeDialog = useCallback((id: DialogId) => {
```

Comment:
Did you mean this?
```suggestion
  const closeDialog = useCallback<CloseDialog>((id: DialogId) => {
```

Reply:
Yes! Though, Typescript should already know the type, as `closeDialog` is part of `value` which again is part of the context which knows `CloseDialog` (check `DialogContextValue`). I wouldn't mind adding it here though even if it's redundant, it communicates stronger what it is about